domain/modelagent/service/service.go

```diff
@@ -499,6 +579,356 @@ func (s *Service) SetUnitReportedAgentVersion(
+// the model running the Juju controller.
+// - [modelagenterrors.ModelUpgradeBlocker] when their exists a blocker in the
+// model that prevents the model from being upgraded.
+func (s *Service) UpgradeModelTargetAgentVersionStreamTo(
+	ctx context.Context,
+	consideredVersion semversion.Number,
+	agentStream modelagent.AgentStream,
+) error {
+	// NOTE (tlm): We don't try and short circuit version upgrading if the model
+	// is already at the current version being considered. This is because this
+	// upgrade also has consider the context of an agent stream change.
+	ctx, span := trace.Start(ctx, trace.NameFromFunc())
+	defer span.End()
+
+	if !agentStream.IsValid() {
+		return errors.New("agent stream is not valid").Add(coreerrors.NotValid)
+	}
+
+	currentTargetVersion, err := s.st.GetModelTargetAgentVersion(ctx)
+	if err != nil {
+		return errors.Errorf(
+			"getting current model target agent version: %w", err,
+		)
+	}
+
+	err = s.validateModelCanBeUpgradedTo(ctx, currentTargetVersion, consideredVersion)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = s.st.SetModelTargetAgentVersionAndStream(
+		ctx, currentTargetVersion, consideredVersion, agentStream,
+	)
+	if err != nil {
+		return errors.Capture(err)
+	}
+	return nil
+}
+
+// validateModelCanBeUpgraded checks if the current model is currently in a
+// state that can have it's target agent version changed/"upgraded". This check
+// does not concern itself with the proposed upgrade version. It only cares
+// about the state of the model and if it suitable to be upgraded.
+//
+// This check always assumes that the caller is upgrading the version and never
+// downgrading. The validation performed is not guaranteed to still be valid
+// after this func returns.
+//
+// The following errors may be expected:
+// - [modelagenterrors.CannotUpgradeControllerModel] when the current model is
+// the model running the Juju controller.
+// - [modelagenterrors.ModelUpgradeBlocker] when their exists a blocker in the
+// model that prevents the model from being upgraded.
+func (s *Service) validateModelCanBeUpgraded(
+	ctx context.Context,
+) error {
+	isControllerModel, err := s.st.IsControllerModel(ctx)
+	if err != nil {
+		return errors.Errorf(
+			"checking if model is considered the controller's model: %w", err,
+		)
+	}
+	if isControllerModel {
+		return errors.New(
+			"model upgrades cannot be performed on the controller model",
+		).Add(modelagenterrors.CannotUpgradeControllerModel)
+	}
+
+	failedMachineCount, err := s.st.GetMachineCountNotUsingBases(ctx, corebase.WorkloadBases())
+	if err != nil {
+		return errors.Errorf(
+			"getting count of machines in model not running a supported workload base: %w",
+			err,
+		)
+	}
+
+	if failedMachineCount > 0 {
+		return modelagenterrors.ModelUpgradeBlocker{
+			Reason: fmt.Sprintf(
+				"model has %d machines not using one of %v supported bases",
+				failedMachineCount, corebase.WorkloadBases(),
```

Comment:
This won't read very well.
Better as `model has %d machines not using one of the supported bases: %v`